### PR TITLE
Gas new features

### PR DESCRIPTION
**Pull request by uniemimu**
Support for telemetry based labels

### CODE REVIEW

File: gpu-aware-scheduling/deploy/gas-deployment.yaml [hunk 1]
```diff
@@ -26,6 +26,13 @@ spec:
         - "--v=4"
         image: intel/gpu-extender
         imagePullPolicy: IfNotPresent
+        securityContext:
```
Comment (togashidm) [inline gpu-aware-scheduling/deploy/gas-deployment.yaml:29]: Nice addition

File: gpu-aware-scheduling/docs/usage.md [hunk 2]
```diff
@@ -65,6 +65,61 @@ Your PODs then, needs to ask for some GPU-resources. Like this:
...
+[Telemetry Aware Scheduling](../../telemetry-aware-scheduling/README.md) labeling strategy,
+which puts them in its own namespace. In practice the supported labels need to be in the
+`telemetry.aware.scheduling.POLICYNAME/` namespace, where the POLICYNAME may be anything.
+
+The node label `gas-deschedule-pods-GPUNAME` where the GPUNAME can be e.g. card0, card1, card2...
+which corresponds to the gpu names under /dev/dri, will result in GAS labeling the PODs which
+use the named GPU with the `gpu.aware.scheduling/deschedule-pod=gpu` label. You may then
+use with a kubernetes descheduler to pick the pods for descheduling. So TAS labels the node, and
+based on the node label GAS finds and labels the PODs. Descheduler can be configured to
+deschedule the pods based on pod labels.
```
Comment (madalazar) [inline gpu-aware-scheduling/docs/usage.md:81]: I would remove this line from all the occurrences below and add it just before L75 in a section called Notes/Abbreviations

GPUNAME can be e.g. card0, card1, card2… which corresponds to the gpu names under /dev/dri

L76-81 I would rewrite as:

The node label `gas-deschedule-pods-GPUNAME`  will result in GAS labeling the PODs which
use the named GPU with the `gpu.aware.scheduling/deschedule-pod=gpu` label. So TAS labels the node, and based on the node label GAS finds and labels the PODs. 
You may then use a kubernetes descheduler to pick the pods for descheduling via their labels.

File: gpu-aware-scheduling/docs/usage.md [hunk 3]
```diff
@@ -65,6 +65,61 @@ Your PODs then, needs to ask for some GPU-resources. Like this:
...
+
+The node label `gas-disable-GPUNAME` where the GPUNAME can be e.g. card0, card1, card2... which
+corresponds to the gpu names under /dev/dri, will result in GAS stopping the use of the named
+GPU for new allocations.
+
+The node label `gas-prefer-gpu=GPUNAME` where the GPUNAME can be e.g. card0, card1, card2...
```
Comment (madalazar) [inline gpu-aware-scheduling/docs/usage.md:87]: Is it gas-prefer-gpu=GPUNAME or gas-prefer-gpu-GPUNAME. Seems like a typo unless you intend to have the gpu name as the label's value.

Comment (uniemimu) [reply gpu-aware-scheduling/docs/usage.md:87]: The intention is to have the gpu name as the label's value. Here the TAS labeling strategy can be used to get the best metrics value producing GPU to have the value field filled up correctly. For that, the label key must match.

File: gpu-aware-scheduling/pkg/gpuscheduler/node_resource_cache.go [hunk 4]
```diff
@@ -123,40 +174,65 @@ func NewCache(client kubernetes.Interface) *Cache {
...
+		}
+	default:
+		return false
+	}
+
+	return hasGPUCapacity(node)
+}
+
 // This must be called with rwmutex unlocked
 // set add=true to add, false to remove resources.
```
[the remark this answers is not in the document]
Comment (uniemimu) [reply gpu-aware-scheduling/pkg/gpuscheduler/node_resource_cache.go:237]: good catch it's adj

**Review by madalazar — APPROVED** [review]
Looks good to me! Approved

File: gpu-aware-scheduling/README.md [hunk 5]
```diff
@@ -32,61 +32,9 @@ A worked example for GAS is available [here](docs/usage.md)
...
-1) Add the ability to get configmaps to the kubeadm scheduler config map. (A cluster role binding for this is at deploy/extender-configuration/configmap-getter.yaml)
-2) Add the ``dnsPolicy: ClusterFirstWithHostNet`` in order to access the scheduler extender by service name.
-
-After these steps the scheduler extender should be registered with the Kubernetes Scheduler.
+You should follow extender configuration instructions from the
+[Telemetry Aware Scheduling](../telemetry-aware-scheduling/README.md#Extender-configuration) and adapt those instructions to
+use GPU Aware Scheduling configurations, which can be found in the [deploy/extender-configuration](deploy/extender-configuration) folder.
 
 #### Deploy GAS
 GPU Aware Scheduling uses go modules. It requires Go 1.13+ with modules enabled in order to build. GAS has been tested with Kubernetes 1.15+.
```
[the remark this answers is not in the document]
Comment (uniemimu) [reply gpu-aware-scheduling/README.md]: done

File: gpu-aware-scheduling/docs/usage.md [hunk 6]
```diff
@@ -65,6 +65,61 @@ Your PODs then, needs to ask for some GPU-resources. Like this:
...
+card1, card2... which corresponds to the gpu names under /dev/dri, the disabling will impact a
+group of GPUs which is defined in the node label `gpu.intel.com/pci-groups`. The syntax of the
+pci group node label is easiest to explain with an example: `gpu.intel.com/pci-groups=0.1_2.3.4`
+would indicate there are two pci-groups in the node separated with an underscore, in which card0
+and card1 form the first group, and card2, card3 and card4 form the second group. If GAS would
+find the node label `gas-disable-card3=PCI_GROUP` in a node with the previous example PCI-group
+label, GAS would stop using card2, card3 and card4 for new allocations, as card3 belongs in that
+group.
+
+`gas-deschedule-pods-GPUNAME` supports the PCI-GROUP value similarly, the whole group in which
```
[the remark this answers is not in the document]
Comment (uniemimu) [reply gpu-aware-scheduling/docs/usage.md]: indeed!

File: gpu-aware-scheduling/pkg/gpuscheduler/node_resource_cache.go [hunk 7]
```diff
@@ -390,53 +556,113 @@ func (c *Cache) deletePodFromCache(podObj interface{}) {
...
+
+func (c *Cache) nodeWorkerRun() {
+	for c.nodeWork() {
+	}
 }
 
-func (c *Cache) workerRun() {
-	for c.work() {
+func (c *Cache) nodeWork() bool {
+	klog.V(L5).Info("node worker started")
```
[the remark this answers is not in the document]
Comment (uniemimu) [reply gpu-aware-scheduling/pkg/gpuscheduler/node_resource_cache.go:604]: This is printed only with the highest verbosity, intentionally, as it prints quite often when enabled. The next alternative is to remove the log print entirely, which is tempting. But leaving for now.

**Review by togashidm — APPROVED** [review]

**Comment by togashidm** [comment]
LGTM

**Pull request merged**